apps/tx-builder/src/components/forms/SolidityForm.tsx

```diff
@@ -0,0 +1,202 @@
+import { useState } from 'react';
+import { SubmitHandler, useForm } from 'react-hook-form';
+import { ButtonLink, Switch, Text } from '@gnosis.pm/safe-react-components';
+import styled from 'styled-components';
+
+import { ContractInterface } from '../../hooks/useServices/interfaceRepository';
+import {
+  ADDRESS_FIELD_TYPE,
+  AMOUNT_FIELD_TYPE,
+  CONTRACT_METHOD_FIELD_TYPE,
+  HEX_ENCODED_DATA_FIELD_TYPE,
+} from './fields/fields';
+import Field from './fields/Field';
+import { Examples } from '../Examples';
+import { encodeToHexData } from '../../utils';
+
+function SolidityForm({
+  id,
+  onSubmit,
+  getAddressFromDomain,
+  initialValues,
+  nativeCurrencySymbol,
+  networkPrefix,
+  contract,
+  children,
+}: SolidityFormPropsTypes) {
+  const [showExamples, setShowExamples] = useState<boolean>(false);
+  const [showHexEncodedData, setShowHexEncodedData] = useState<boolean>(false);
+
+  const { handleSubmit, control, setValue, watch, getValues, reset, clearErrors } = useForm<SolodityFormValues>({
+    defaultValues: initialValues,
+    mode: 'onTouched', // This option allows you to configure the validation strategy before the user submits the form
+  });
+
+  const contractMethodIndex = watch(CONTRACT_METHOD_INDEX_FIELD_NAME);
```

Comment:
If you use [`useWatch`](https://react-hook-form.com/api/usewatch), it will isolate re-renders.

Reply:
mmm, I think that in this case we should use `watch()` right?

I think that [useWatch](https://react-hook-form.com/api/usewatch) its meant to be used in a nested component. useWatch will isolate re-render only that component and potentially result in better performance.

Also that component needs to be wrapped with a [FormProvider](https://react-hook-form.com/advanced-usage#FormProviderPerformance), it uses a [React's Context](https://reactjs.org/docs/context.html) API. 
It think that it only makes sense if the data is passed through a very complex components tree. In that case it saves you to having to pass props down manually at every level.

What is your opinion about that?

Reply:
I suppose as you're using it within this component then it may not be necessary.

It requires `control`, which is likely the explanation for the context.